text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1131 @@
+| Module                                          | Global                   |
+| ---                                             | ---                      |
+| `import { begin } from "@ember/runloop"`        | `Ember.run.begin`        |
+| `import { bind } from "@ember/runloop"`         | `Ember.run.bind`         |
+| `import { cancel } from "@ember/runloop"`       | `Ember.run.cancel`       |
+| `import { debounce } from "@ember/runloop"`     | `Ember.run.debounce`     |
+| `import { end } from "@ember/runloop"`          | `Ember.run.end`          |
+| `import { join } from "@ember/runloop"`         | `Ember.run.join`         |
+| `import { later } from "@ember/runloop"`        | `Ember.run.later`        |
+| `import { next } from "@ember/runloop"`         | `Ember.run.next`         |
+| `import { once } from "@ember/runloop"`         | `Ember.run.once`         |
+| `import { run } from "@ember/runloop"`          | `Ember.run`              |
+| `import { schedule } from "@ember/runloop"`     | `Ember.run.schedule`     |
+| `import { scheduleOnce } from "@ember/runloop"` | `Ember.run.scheduleOnce` |
+| `import { throttle } from "@ember/runloop"`     | `Ember.run.throttle`     |
+
+### `@ember/service`
+| Module                                    | Global                 |
+| ---                                       | ---                    |
+| `import Service from "@ember/service"`    | `Ember.Service`        |
+| `import { inject } from "@ember/service"` | `Ember.inject.service` |
+
+### `@ember/string`
+| Module                                       | Global                    |
+| ---                                          | ---                       |
+| `import { camelize } from "@ember/string"`   | `Ember.String.camelize`   |
+| `import { capitalize } from "@ember/string"` | `Ember.String.capitalize` |
+| `import { classify } from "@ember/string"`   | `Ember.String.classify`   |
+| `import { dasherize } from "@ember/string"`  | `Ember.String.dasherize`  |
+| `import { decamelize } from "@ember/string"` | `Ember.String.decamelize` |
+| `import { fmt } from "@ember/string"`        | `Ember.String.fmt`        |
+| `import { htmlSafe } from "@ember/string"`   | `Ember.String.htmlSafe`   |
+| `import { loc } from "@ember/string"`        | `Ember.String.loc`        |
+| `import { underscore } from "@ember/string"` | `Ember.String.underscore` |
+| `import { w } from "@ember/string"`          | `Ember.String.w`          |
+
+### `@ember/utils`
+| Module                                     | Global            |
+| ---                                        | ---               |
+| `import { isBlank } from "@ember/utils"`   | `Ember.isBlank`   |
+| `import { isEmpty } from "@ember/utils"`   | `Ember.isEmpty`   |
+| `import { isNone } from "@ember/utils"`    | `Ember.isNone`    |
+| `import { isPresent } from "@ember/utils"` | `Ember.isPresent` |
+| `import { tryInvoke } from "@ember/utils"` | `Ember.tryInvoke` |
+| `import { type
```

Comment:
@les2 Thank you for bringing this up. I discussed it with the core team, and we agree wholeheartedly with your concerns but believe they are addressable even with the approach proposed in this RFC.

First, we should make it clear that while it's not possible to import JavaScript modules conditionally _at runtime_, it is definitely possible to conditionally include or remove modules at build time via a Broccoli plugin. We also expect TC39 to add additional syntax for conditional imports in the future, since this is a general problem with imports that isn't specific to prototype extensions.

Second, there are a few downsides to exporting a function that installs prototype extensions when called.

The first problem is that, of course, you still have to parse, construct, and export the function even if no one ever calls it. Conversely, a module that never gets imported would be trivially removed by any modern module bundler.

The second problem, which I think is even worse, is that once you provide the function it can be called at any time. If some code requires the function and waits for some async behavior to complete before calling it, you could end up in a nightmare scenario where the time at which the prototype extensions are applied is non-deterministic.

> Prototype extensions are currently controlled by the environment settings. Where would this import go? The app module?

I think you've hit the nail on the head here. I think what we can do is look at the configuration and have the Ember CLI build generate the appropriate `import` statement if prototype extensions are enabled. This makes the feature more discoverable for end users, retains the ability to have different builds with it either enabled or disabled, still provides a path for non-Ember CLI users to import the prototype extensions if they wish, and ensures that the prototype extension happens at a predictable place and time (i.e., not at runtime).

At some point, I would like for us to disable prototype extensions by default, and then these modules would stay around just for compatibility reasons anyway. The future is in clean prototypes.

> Could an addon import this module and enable proto extensions for the entire app?

Yes, but this is true today as well; nothing stops an addon from extending prototypes however it wants.